doc: add the upgrade guide from 5.2 to 5.3 (#13960)
scylladb/scylladb: docs/upgrade/upgrade-opensource/upgrade-guide-from-5.2-to-5.3/upgrade-guide-from-5.2-to-5.3-generic.rst

.. |SCYLLA_NAME| replace:: ScyllaDB

.. |SRC_VERSION| replace:: 5.2
.. |NEW_VERSION| replace:: 5.3

.. |DEBIAN_SRC_REPO| replace:: Debian
.. _DEBIAN_SRC_REPO: https://www.scylladb.com/download/?platform=debian-10&version=scylla-5.2

.. |UBUNTU_SRC_REPO| replace:: Ubuntu
.. _UBUNTU_SRC_REPO: https://www.scylladb.com/download/?platform=ubuntu-20.04&version=scylla-5.2

.. |SCYLLA_DEB_SRC_REPO| replace:: ScyllaDB deb repo (|DEBIAN_SRC_REPO|_, |UBUNTU_SRC_REPO|_)

.. |SCYLLA_RPM_SRC_REPO| replace:: ScyllaDB rpm repo
.. _SCYLLA_RPM_SRC_REPO: https://www.scylladb.com/download/?platform=centos&version=scylla-5.2

.. |DEBIAN_NEW_REPO| replace:: Debian

> .. _DEBIAN_NEW_REPO: https://www.scylladb.com/download/?platform=debian-10&version=scylla-5.2

Reviewer: 5.3?

.. _SCYLLA_METRICS: ../metric-update-5.2-to-5.3

=============================================================================
Upgrade Guide - |SCYLLA_NAME| |SRC_VERSION| to |NEW_VERSION|
=============================================================================

This document is a step-by-step procedure for upgrading from |SCYLLA_NAME| |SRC_VERSION| to |SCYLLA_NAME| |NEW_VERSION|, and rollback to 
version |SRC_VERSION| if required.

This guide covers upgrading ScyllaDB on Red Hat Enterprise Linux (RHEL), CentOS, Debian, and Ubuntu. See 
:doc:`OS Support by Platform and Version </getting-started/os-support>` for information about supported versions.

It also applies when using ScyllaDB official image on EC2, GCP, or Azure. The image is based on Ubuntu 22.04.

Upgrade Procedure
=================

A ScyllaDB upgrade is a rolling procedure that does **not** require full cluster shutdown.
For each of the nodes in the cluster, serially (i.e., one node at a time), you will:

* Check that the cluster's schema is synchronized
* Drain the node and backup the data
* Backup the configuration file
* Stop ScyllaDB
* Download and install new ScyllaDB packages

> * (Optional) Enable consistent cluster management in the configuration file

Reviewer: s/Enable/Disable

            The file is included in the ScyllaDB packages downloaded in the previous step. The file location is ``http://downloads.scylladb.com/downloads/scylla/aws/manifest/scylla-packages-<version>-<arch>.txt``

            Example:

                .. code:: sh

                   cat scylla-packages-5.2.0-x86_64.txt | sudo xargs -n1 apt-get install -y

                .. note::

                   Alternatively, you can update the manifest file with the following command:

                   ``sudo apt-get install $(awk '{print $1'} scylla-packages-<version>-<arch>.txt) -y``

        .. _upgrade-image-upgrade-guide-regular-procedure:

        **To upgrade ScyllaDB:**

        #. Update the |SCYLLA_DEB_NEW_REPO| to |NEW_VERSION|.

        #. Install the new ScyllaDB version:

            .. code-block:: console

               sudo apt-get clean all
               sudo apt-get update
               sudo apt-get dist-upgrade scylla

        Answer ‘y’ to the first two questions.

(Optional) Disable consistent cluster management in the node's configuration file
-----------------------------------------------------------------------------------

> In ScyllaDB 5.3. Raft-based consistent cluster management is enabled by default. This means that Raft will be enabled and used 

Reviewer: `In ScyllaDB 5.3.` -> `In ScyllaDB 5.3,`

> In ScyllaDB 5.3. Raft-based consistent cluster management is enabled by default. This means that Raft will be enabled and used 

Reviewer: BTW. looking at the code, this is not true. It's disabled by default.
It is only enabled by default in new clusters which are using a fresh conf/scylla.yaml file (that's true already in 5.2).
But for old clusters which don't explicitly state the `consistent_cluster_management` option in conf/scylla.yaml, it is disabled.
```cpp
     , consistent_cluster_management(this, "consistent_cluster_management", value_status::Used, false, "Use RAFT for cluster management and DDL")
```
@kostja @mykaul do we want to change it?

to consistently manage cluster-wide metadata as soon as you finish upgrading every node to the new version. 
See the :doc:`Raft in ScyllaDB document </architecture/raft/>` to learn more.

> In ScyllaDB 5.3, you can explicitly disable Raft-based consistent cluster management if needed. 

Reviewer: We should add:
this is only possible if you haven't already enabled consistent cluster management on all nodes before.

In further versions, the option will be removed, and consistent cluster management will be enabled unconditionally.

To disable the option, modify the ``scylla.yaml`` configuration file in ``/etc/scylla/`` and set the following parameter to ``false``:

.. code:: yaml

    consistent_cluster_management: false

Start the node
--------------

.. code:: sh

   sudo service scylla-server start

Validate
--------
#. Check cluster status with ``nodetool status`` and make sure **all** nodes, including the one you just upgraded, are in ``UN`` status.
#. Use ``curl -X GET "http://localhost:10000/storage_service/scylla_release_version"`` to check the ScyllaDB version. Validate that the version matches the one you upgraded to.
#. Check scylla-server log (by ``journalctl _COMM=scylla``) and ``/var/log/syslog`` to validate there are no new errors in the log.
#. Check again after two minutes, to validate no new issues are introduced.

Once you are sure the node upgrade was successful, move to the next node in the cluster.

See |Scylla_METRICS|_ for more information..

After upgrading every node
==========================

> The following section applies only if the ``consistent_cluster_management`` option is enabled on every node when upgrading the cluster.

Reviewer: and it wasn't enabled before

Rollback Steps
==============
Drain and gracefully stop the node
----------------------------------

.. code:: sh

   nodetool drain
   sudo service scylla-server stop

Download and install the old release
------------------------------------

..
    TODO: downgrade for 3rd party packages in EC2/GCP/Azure - like in the upgrade section?

.. tabs::

   .. group-tab:: Debian/Ubuntu

        #. Remove the old repo file.

            .. code:: sh

               sudo rm -rf /etc/apt/sources.list.d/scylla.list

        #. Update the |SCYLLA_DEB_SRC_REPO| to |SRC_VERSION|.
        #. Install:

            .. code-block::

               sudo apt-get update
               sudo apt-get remove scylla\* -y
               sudo apt-get install scylla

        Answer ‘y’ to the first two questions.

   .. group-tab:: RHEL/CentOS

        #. Remove the old repo file.

            .. code:: sh

               sudo rm -rf /etc/yum.repos.d/scylla.repo

        #. Update the |SCYLLA_RPM_SRC_REPO|_  to |SRC_VERSION|.
        #. Install:

            .. code:: console

               sudo yum clean all
               sudo rm -rf /var/cache/yum

>                sudo yum remove scylla\\*tools-core

Reviewer 2: sudo yum remove scylla\\*cqlsh

>                sudo yum remove scylla\\*tools-core

Reviewer 2: ```suggestion
               sudo yum remove scylla\*cqlsh
```

Reviewer 2: this is the only change needed for the downgrade instructions 
`scylla-tools-core` is no longer "new" package that conflicts with older versions, `scylla-cqlsh` is a new package that conflicts with older versions of `scylla-tools-java`

.. _SCYLLA_METRICS: ../metric-update-5.2-to-5.3

=============================================================================
Upgrade Guide - |SCYLLA_NAME| |SRC_VERSION| to |NEW_VERSION|
=============================================================================

This document is a step-by-step procedure for upgrading from |SCYLLA_NAME| |SRC_VERSION| to |SCYLLA_NAME| |NEW_VERSION|, and rollback to 
version |SRC_VERSION| if required.

This guide covers upgrading ScyllaDB on Red Hat Enterprise Linux (RHEL), CentOS, Debian, and Ubuntu. See 
:doc:`OS Support by Platform and Version </getting-started/os-support>` for information about supported versions.

It also applies when using ScyllaDB official image on EC2, GCP, or Azure. The image is based on Ubuntu 22.04.

Upgrade Procedure
=================

A ScyllaDB upgrade is a rolling procedure that does **not** require full cluster shutdown.
For each of the nodes in the cluster, serially (i.e., one node at a time), you will:

* Check that the cluster's schema is synchronized
* Drain the node and backup the data
* Backup the configuration file
* Stop ScyllaDB
* Download and install new ScyllaDB packages

> * (Optional) Enable consistent cluster management in the configuration file

Reviewer 3: I think in 5.3 we'll do this step automatically - see https://github.com/scylladb/scylladb/pull/13984 (not done yet though!)

* Start ScyllaDB
* Validate that the upgrade was successful

Apply the following procedure **serially** on each node. Do not move to the next node before validating that the node you upgraded is up and running the new version.

**During** the rolling upgrade, it is highly recommended:

* Not to use the new |NEW_VERSION| features.

> * Not to run administration functions, like repairs, refresh, rebuild, or add or remove nodes. See `sctool <https://manager.docs.scylladb.com/stable/sctool/>`_ for suspending ScyllaDB Manager (only available for ScyllaDB Enterprise) scheduled or running repairs.

Reviewer 3: ```suggestion
* Not to run administration functions, such as repairs, refresh, rebuild, or add or remove nodes. See `sctool <https://manager.docs.scylladb.com/stable/sctool/>`_ for suspending ScyllaDB Manager (only available for ScyllaDB Enterprise) scheduled or running repairs.
```

* Not to apply schema changes.

.. note:: 

>    If you use the `ScyllaDB Monitoring Stack <https://monitoring.docs.scylladb.com/>`_, we recommend upgrading the Monitoring Stack  to the latest version **before** upgrading ScyllaDB.

Reviewer 3: ```suggestion
   If you use the `ScyllaDB Monitoring Stack <https://monitoring.docs.scylladb.com/>`_, we recommend upgrading the Monitoring Stack to the latest version **before** upgrading ScyllaDB.
```

Upgrade Steps
=============

Check the cluster schema
-------------------------

Make sure that all nodes have the schema synchronized before the upgrade. The upgrade procedure will fail if there is a schema 
disagreement between nodes.

.. code:: sh

   nodetool describecluster

Drain the nodes and backup data
-----------------------------------

Before any major procedure, like an upgrade, it is recommended to backup all the data to an external device. In ScyllaDB, 
backup is performed using the ``nodetool snapshot`` command. For **each** node in the cluster, run the following command:

.. code:: sh

   nodetool drain
   nodetool snapshot

Take note of the directory name that nodetool gives you, and copy all the directories having that name under ``/var/lib/scylla`` 

> to a backup device.

Reviewer 3: ```suggestion
to an external backup device.
```

When the upgrade is completed on all nodes, remove the snapshot with the ``nodetool clearsnapshot -t <snapshot>`` command to prevent 
running out of space.

Backup the configuration file
------------------------------
.. code:: sh

   sudo cp -a /etc/scylla/scylla.yaml /etc/scylla/scylla.yaml.backup-src

Gracefully stop the node
------------------------

.. code:: sh

>    sudo service scylla-server stop

Author: I was explicitly told to consistently use "service" across the docs. If you think something else should be used, please open a separate issue to start the discussion. After that, I'll need to update all the occurrences.

               sudo apt-get clean all
               sudo apt-get update
               sudo apt-get dist-upgrade scylla

        Answer ‘y’ to the first two questions.

   .. group-tab:: RHEL/CentOS

        Before upgrading, check what version you are running now using ``scylla --version``. You should use the same version as this version in case you want to |ROLLBACK|_ the upgrade. If you are not running a |SRC_VERSION|.x version, stop right here! This guide only covers |SRC_VERSION|.x to |NEW_VERSION|.y upgrades.

        **To upgrade ScyllaDB:**

        #. Update the |SCYLLA_RPM_NEW_REPO|_  to |NEW_VERSION|.
        #. Install the new ScyllaDB version:

            .. code:: sh

               sudo yum clean all
               sudo yum update scylla\* -y

   .. group-tab:: EC2/GCP/Azure Ubuntu Image

        Before upgrading, check what version you are running now using ``scylla --version``. You should use the same version as this version in case you want to |ROLLBACK|_ the upgrade. If you are not running a |SRC_VERSION|.x version, stop right here! This guide only covers |SRC_VERSION|.x to |NEW_VERSION|.y upgrades.

        There are two alternative upgrade procedures:

>         * :ref:`Upgrading ScyllaDB and simultaneously updating 3rd party and OS packages <upgrade-image-recommended-procedure>`. It is recommended if you are running a ScyllaDB official image (EC2 AMI, GCP, and Azure images), which is based on Ubuntu 20.04.

Reviewer 3: Not Ubuntu 22.04?

Author: You're right; I didn't update it here by mistake. This case shows the weakness of adding the same information in multiple places and maintaining it separately. In the future update, I'll remove the information from this paragraph, as we have it in another place in the docs, and it's not super relevant here.

        .. _upgrade-image-upgrade-guide-regular-procedure:

        **To upgrade ScyllaDB:**

        #. Update the |SCYLLA_DEB_NEW_REPO| to |NEW_VERSION|.

        #. Install the new ScyllaDB version:

            .. code-block:: console

               sudo apt-get clean all
               sudo apt-get update
               sudo apt-get dist-upgrade scylla

        Answer ‘y’ to the first two questions.

.. _upgrade-5.3-enable-raft:

(Optional) Enable consistent cluster management in the node's configuration file
--------------------------------------------------------------------------------

.. note::

   Skip this section if you enabled Raft-based consistent cluster management in version 5.2.

Optionally, you can enable Raft-based consistent cluster management on your cluster with the ``consistent_cluster_management``
option. Setting the option to ``true`` on every node enables the Raft consensus algorithm for your cluster. Raft will be used
to consistently manage cluster-wide metadata as soon as you finish upgrading every node to the new version. See 
:doc:`Raft in ScyllaDB </architecture/raft/>` to learn more.

> In ScyllaDB 5.2 and 5.3, Raft-based consistent cluster management is disabled by default. If you didn't enable the feature

Reviewer 3: I thought for new deployments it's enabled by default?

Reviewer: What Yaniv meant is that in 5.2/5.3, it's enabled by default in **new deployments** i.e. when you start a completely new cluster; it's not yet enabled by default during upgrade (this is what the decision that needs to be made is about).

Indeed, we should probably say something like (but this can be fixed in a follow-up - I already queued the patch)
"When you upgrade to ScyllaDB 5.2 and 5.3, Raft-based consistent cluster management is disabled by default"

- to distinguish from the situation of starting a new cluster (where it will be enabled by default because of provided conf/scylla.yaml which enables it)

Author: Thanks, I get it now. You're both right - I'll clarify the information in the follow-up PR.
